log/sync.go

```diff
@@ -0,0 +1,195 @@
+func NewSyncWriter(w io.Writer) *SyncWriter {
+	return &SyncWriter{w: w}
+}
+
+// Write writes p to the underlying io.Writer. If another write is already in
+// progress, the calling goroutine blocks until the SyncWriter is available.
+func (w *SyncWriter) Write(p []byte) (n int, err error) {
+	w.mu.Lock()
+	n, err = w.w.Write(p)
+	w.mu.Unlock()
+	return n, err
+}
+
+// syncLogger provides concurrent safe logging for another Logger.
+type syncLogger struct {
+	mu     sync.Mutex
+	logger Logger
+}
+
+// NewSyncLogger returns a logger that synchronizes concurrent use of the
+// wrapped logger. When multiple goroutines use the SyncLogger concurrently
+// only one goroutine will be allowed to log to the wrapped logger at a time.
+// The other goroutines will block until the logger is available.
+func NewSyncLogger(logger Logger) Logger {
+	return &syncLogger{logger: logger}
+}
+
+// Log logs keyvals to the underlying Logger. If another log is already in
+// progress, the calling goroutine blocks until the syncLogger is available.
+func (l *syncLogger) Log(keyvals ...interface{}) error {
+	l.mu.Lock()
+	err := l.logger.Log(keyvals...)
+	l.mu.Unlock()
+	return err
+}
+
+// AsyncLogger provides buffered asynchronous and concurrent safe logging for
+// another logger.
+//
+// If the wrapped logger's Log method ever returns an error, the AsyncLogger
+// will stop processing log events and make the error available via the Err
+// method. Any unprocessed log events in the buffer will be lost.
+type AsyncLogger struct {
+	logger   Logger
+	keyvalsC chan []interface{}
+
+	stopping chan struct{}
+	stopped  chan struct{}
+
+	mu  sync.Mutex
+	err error
+}
+
+// NewAsyncLogger returns a new AsyncLogger that logs to logger and can buffer
+// up to size log events before overflowing.
+func NewAsyncLogger(logger Logger, size int) *AsyncLogger {
+	l := &AsyncLogger{
+		logger:   logger,
+		keyvalsC: make(chan []interface{}, size),
+		stopping: make(chan struct{}),
+		stopped:  make(chan struct{}),
+	}
+	go l.run()
+	return l
+}
+
+// run forwards log events from l.keyvalsC to l.logger.
+func (l *AsyncLogger) run() {
+	defer close(l.stopped)
+	for keyvals := range l.keyvalsC {
+		err := l.logger.Log(keyvals...)
+		if err != nil {
+			l.mu.Lock()
+			l.stop(err)
+			l.mu.Unlock()
+			return
+		}
+	}
+}
+
+// caller must hold l.mu
+func (l *AsyncLogger) stop(err error) {
+	if err != nil && l.err == nil {
+		l.err = err
+	}
+	select {
+	case <-l.stopping:
+		// already stopping, do nothing
+	default:
+		close(l.stopping)
+		close(l.keyvalsC)
+	}
+}
+
+// Log queues keyvals for logging by the wrapped Logger. Log may be called
+// concurrently by multiple goroutines. If the the buffer is full, Log will
+// return ErrAsyncLoggerOverflow and the keyvals are not queued. If the
+// AsyncLogger is stopping, Log will return ErrAsyncLoggerStopping.
+func (l *AsyncLogger) Log(keyvals ...interface{}) error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
```

Comment:
This is interesting. At first I thought it was an oversight, but now I see it's to prevent a race. Still, it seems a shame to require a mutex op with every Log; my intuition is that there's a clever way to avoid it...